dbms/src/Server/StorageConfigParser.cpp

```diff
@@ -444,9 +446,129 @@ std::tuple<size_t, TiFlashStorageConfig> TiFlashStorageConfig::parseSettings(
         storage_config.remote_cache_config.parse(config.getString("storage.remote.cache"), log);
     }
 
+    if (config.has("storage.tmp"))
+    {
+        // Need to make sure storage.main/latest is parsed before storage.tmp.
+        storage_config.parseAndCreateTmpPath(config.getString("storage"), global_capacity_quota, log);
+    }
+    else if (config.has("tmp_path"))
+    {
+        storage_config.tmp_path = getNormalizedPath(config.getString("tmp_path"));
+        // If storage.tmp doesn't exist, tmp_capacity will be zero.
+        storage_config.tmp_capacity = 0;
+    }
+    else
+    {
+        storage_config.tmp_path = storage_config.latest_data_paths[0] + "tmp/";
+        // If storage.tmp doesn't exist, tmp_capacity will be zero.
+        storage_config.tmp_capacity = 0;
+    }
+
     return std::make_tuple(global_capacity_quota, storage_config);
 }
 
+void TiFlashStorageConfig::parseAndCreateTmpPath(
+    const String & content,
+    UInt64 global_capacity_quota,
+    const LoggerPtr & log)
+{
+    // global_capacity_quota and storage.main/latest.capacity cannot take effects at the same time.
+    RUNTIME_CHECK(!(!main_capacity_quota.empty() && global_capacity_quota > 0));
+    RUNTIME_CHECK_MSG(
+        (main_capacity_quota.size() == latest_capacity_quota.size()) && !main_capacity_quota.empty(),
+        "main_capacity_quota.size: {}, latest_capacity_quota.size: {}",
+        main_capacity_quota.size(),
+        latest_capacity_quota.size());
+
+    std::istringstream ss(content);
+    cpptoml::parser p(ss);
+    auto table = p.parse();
+
+    auto tmp_path_opt = table->get_qualified_as<String>("tmp.dir");
+    if (!tmp_path_opt || tmp_path_opt->empty())
+        tmp_path = latest_data_paths[0] + "tmp/";
+    else
+        tmp_path = *tmp_path_opt;
+
+    tmp_path = getNormalizedPath(tmp_path);
```

Comment:
Add an info logging about the value of `tmp_path` here?

Reply:
Added in L465